[libc++] [libc++abi] Initialize exception directly in make_exception_ptr if __cxa_init_primary_exception is available in ABI-library

itrofimow opened the pull request:
This patch implements an ABI-extension function `__cxa_init_primary_exception` in libcxxabi, which is already present as an extension in both libsupc++ and libcxxrt, and makes use of this function in `std::make_exception_ptr`: now instead of going through full throw + catch cycle we are able to initialize an exception directly, thus making `std::make_exception_ptr` around x30 times faster.

itrofimow committed:
[libcxx] [libcxxabi] don't throw+catch in std::make_exception_ptr, if possible

Differential Revision: https://reviews.llvm.org/D158620

itrofimow committed:
+ clang-format

itrofimow committed:
cleanup

itrofimow committed:
+ clang-format, fix libcxx.imp

itrofimow committed:
wip on making the build green

itrofimow committed:
wip on making the build green

itrofimow committed:
bits of formatting

itrofimow committed:
ABI fixes

itrofimow committed:
rebase fix

itrofimow committed:
rebase fix

itrofimow committed:
+ clang-format

itrofimow committed:
+ generate-cxx-abilist

itrofimow committed:
fix compilcation under gcc

itrofimow committed:
some more ifdefs

itrofimow committed:
+ clang-format

itrofimow committed:
+ clang-format

itrofimow committed:
+ clang-format, 17.0.2

itrofimow commented:
I had to split `typeinfo` header to break the `exception` -> `typeinfo` -> `exception` circular dependency, hence a sizeable +/-, other than that the PR is basically a couple dozen LoC.

The build is failing mostly due to missing changes in abi-lists for various architectures, i'm planning to fix that once/if the general idea is approved.

One question left is the way to determine whether the new `__cxa_init_primary_exception` function is available: right now i'm relying on `LIBCXX_BUILDING_LIBCXXABI`, but i think it's not enough. 
Should i add a new define into libcxxabi, or check for `__cxa_init_primary_exception` presence at runtime via `dlsym`?
Personally i'm leaning towards `dlsym`, since it would not only allow linking with other versions of libcxxabi, but potentially implementing the same optimization when linked against libcxxrt, see https://github.com/libcxxrt/libcxxrt/pull/23.

@philnik777 @EricWF Could you please have a look?

philnik777 commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -12,8 +12,10 @@
+#include <__typeinfo/typeinfo.h>
 #include <cstddef>
 #include <cstdlib>
+#include <type_traits>
```
Please include the required detail headers instead.

philnik777 commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -51,11 +69,24 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
 _LIBCPP_HIDE_FROM_ABI exception_ptr make_exception_ptr(_Ep __e) _NOEXCEPT {
 #  ifndef _LIBCPP_HAS_NO_EXCEPTIONS
+#    if defined(LIBCXX_BUILDING_LIBCXXABI)
+  using _Ep2 = typename decay<_Ep>::type;
```
```suggestion
  using _Ep2 = __decay_t<_Ep>;
```

philnik777 commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -51,11 +69,24 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
 template <class _Ep>
 _LIBCPP_HIDE_FROM_ABI exception_ptr make_exception_ptr(_Ep __e) _NOEXCEPT {
 #  ifndef _LIBCPP_HAS_NO_EXCEPTIONS
+#    if defined(LIBCXX_BUILDING_LIBCXXABI)
```
This should be guarded by an availability macro instead.

itrofimow replied:
Should it be a macro or a runtime availability check instead?

itrofimow replied:
I've added support for `libcxxrt` as well and made it a runtime check

itrofimow replied:
The `ifdef` became a bit too big to copy-paste it, so i moved it into `__config`

philnik777 commented on libcxx/include/typeinfo:
```diff
@@ -57,365 +57,13 @@ public:
 */
 
```
Let's do the header split i it's own PR instead.

itrofimow replied:
https://github.com/llvm/llvm-project/pull/68812

itrofimow replied:
We decided to drop backwards compatibility include (typeinfo -> exception) instead

philnik777 commented on libcxxabi/include/cxxabi.h:
```diff
@@ -36,13 +36,19 @@ class type_info; // forward declaration
 __cxa_free_exception(void *thrown_exception) throw();
+// This function is an llvm extension
+extern _LIBCXXABI_FUNC_VIS __cxa_exception*
+__cxa_init_primary_exception(void* object, std::type_info* tinfo, void(_LIBCXXABI_DTOR_FUNC* dest)(void*)) throw();
```
This probably requires availability annotations. CC @ldionne

ldionne replied:
Actually, I would like to have the opinion of some other folks on this first. We generally try to keep libc++abi close to the specification and I'm really unsure about adding an "LLVM extension".

@rjmccall @MaskRay Do you have an opinion about this patch? I am a bit wary of adding extensions to the Itanium C++ ABI in libc++abi.

itrofimow replied:
To add some context to the discussion:

- gcc has this function as "GNU extension": https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/libsupc%2B%2B/cxxabi_init_exception.h#L70, which was added with the exact same performance reasoning in mind: https://gcc.gnu.org/legacy-ml/libstdc++/2016-08/msg00036.html
- as of recently `libcxxrt` has this function as well: https://github.com/libcxxrt/libcxxrt/blob/master/src/cxxabi.h#L211

jyknight replied:
Coordinating to avoid divergence/ABI incompatibility with GCC is important, so that we're implementing a pre-existing GCC ABI extension is very important to mention!

The context that this is an extension that GCC already implements, and that LLVM is now implementing compatibly, should to be mentioned in the PR description and in the comments for this function. (Presuming that our implementation _is_ ABI-compatible -- which it of course needs to be!)

Given that, IMO, it's fine to make this ABI extension.

ldionne replied:
Yeah, I think that changes everything indeed.

EricWF replied:
> * as of recently `libcxxrt` has this function as well: [https://github.com/libcxxrt/libcxxrt/blob/master/src/cxx](https://github.com/libcxxrt/libcxxrt/blob/master/src/cxxabi.h#L211)

It's worth noting that you added the libcxxrt extension.

itrofimow replied:
> It's worth noting that you added the libcxxrt extension.

Agreed, didn't mean to try to trick anyone.

The point i'm trying to make is that
a) both libsupc++ and libcxxrt found the extension reasonable
b) libcxx could make use of the extension already present in aforementioned implementation, and it wouldn't be unreasonable to add it into libcxxabi as well

philnik777 reviewed (changes requested):
I haven't looked at the actual changes closely yet, but the amount of extra code seems reasonable for a significant speedup.

itrofimow committed:
code review fixes

itrofimow committed:
fix transitive_includes tests

itrofimow committed:
runtime lookup of __cxa_init_primary_exception

itrofimow committed:
regenerate cxx-abilist

itrofimow committed:
add exception_ptr.bench

itrofimow committed:
add threaded benchmark

itrofimow committed:
better measurements in benchmark

itrofimow committed:
add cxx-abilist for other architectures, some ifdefs for windows

itrofimow committed:
support libcxxrt as well

itrofimow committed:
add remaining cxx-abilists

itrofimow committed:
fix the build for older versions of libcxxrt, add some comments

itrofimow committed:
cleaner benchmark

itrofimow commented:
Benchmarks to back the performance claims up:
```
itrofimow@itrofimow-nix:~/work/llvm-project/ex_ptr_bench$ taskset -c 8-15 ~/work/gbench/tools/compare.py benchmarks ./main_libcxxabi ./patch_libcxxabi 
RUNNING: ./main_libcxxabi --benchmark_out=/tmp/tmpi96tlrpa
2023-10-15T20:57:14+03:00
Running ./main_libcxxabi
Run on (16 X 1280.96 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)
Load Average: 0.36, 1.30, 1.61
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
bm_make_exception_ptr/threads:1        745 ns          745 ns       950011
bm_make_exception_ptr/threads:2        816 ns         1633 ns       455610
bm_make_exception_ptr/threads:4        863 ns         3208 ns       227044
bm_make_exception_ptr/threads:8       1120 ns         8064 ns        87896
RUNNING: ./patch_libcxxabi --benchmark_out=/tmp/tmpvd2nklsl
2023-10-15T20:57:16+03:00
Running ./patch_libcxxabi
Run on (16 X 1259.78 MHz CPU s)
CPU Caches:
  L1 Data 48 KiB (x8)
  L1 Instruction 32 KiB (x8)
  L2 Unified 1280 KiB (x8)
  L3 Unified 24576 KiB (x1)
Load Average: 0.36, 1.30, 1.61
--------------------------------------------------------------------------
Benchmark                                Time             CPU   Iterations
--------------------------------------------------------------------------
bm_make_exception_ptr/threads:1       25.6 ns         25.6 ns     27321613
bm_make_exception_ptr/threads:2       12.7 ns         25.4 ns     27881074
bm_make_exception_ptr/threads:4       6.84 ns         27.4 ns     27500684
bm_make_exception_ptr/threads:8       3.56 ns         28.5 ns     25787520
Comparing ./main_libcxxabi to ./patch_libcxxabi
Benchmark                                         Time             CPU      Time Old      Time New       CPU Old       CPU New
------------------------------------------------------------------------------------------------------------------------------
bm_make_exception_ptr/threads:1                -0.9656         -0.9656           745            26           745            26
bm_make_exception_ptr/threads:2                -0.9844         -0.9844           816            13          1633            25
bm_make_exception_ptr/threads:4                -0.9921         -0.9915           863             7          3208            27
bm_make_exception_ptr/threads:8                -0.9968         -0.9965          1120             4          8064            28
OVERALL_GEOMEAN                                -0.9892         -0.9887             0             0             0             0
```

It can be seen that with the patch not only the performance is x30(!) in single-threaded case, but it also scales linearly with the number of threads (CPU / Time == number of threads), when on current thunk it's not the case: i'm using not the most recent gcc's unwinder, so contention on the lock in `dl_iterate_phdr` takes place. 

The effect is the same with `libcxxrt`, although it seems to be a bit slower overall.

I've addressed review comments and the build is green. so re-requesting the review.

jyknight commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -26,6 +28,22 @@ namespace std { // purposefully not using versioning namespace
   void* __ptr_;
 
+#  ifndef _LIBCPP_HAS_NO_EXCEPTIONS
+#    if defined(LIBCXXRT) || defined(LIBCXX_BUILDING_LIBCXXABI)
```
Why does this need to be conditional on libcxxrt or libcxxabi, if libstdc++/libsupc++ also implements this ABI?

itrofimow replied:
Initially i didn't want to do that due to some amount of copy-paste, but you're right, why not

jyknight commented on libcxx/src/support/runtime/exception_pointer_cxxabi.ipp:
```diff
@@ -34,6 +48,44 @@ exception_ptr& exception_ptr::operator=(const exception_ptr& other) noexcept
+    // and the caller knows how to handle this (it fallbacks to throw + catch).
+    using CxaInitPrimaryExceptionPrototype = void *(*)(void *, type_info *, void(*)(void *));
+    static CxaInitPrimaryExceptionPrototype cxa_init_primary_exception_fn = reinterpret_cast<CxaInitPrimaryExceptionPrototype>(
+        dlsym(RTLD_DEFAULT, "__cxa_init_primary_exception"));
```
It would be better to use a weak reference to the __cxa_init_primary_exception instead of dlsym.

That way, the loader takes care of the lookup, and if you are building libc++ and libc++abi together (which is the common case), there's no extra overhead.

itrofimow replied:
Apparently that doesn't seem to work with Apple.
I'm fine with that, since i'm only interested in linux platform, and for now i ifdefed-out Apple

jyknight commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -51,11 +69,34 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
+#    if defined(LIBCXXRT) || defined(LIBCXX_BUILDING_LIBCXXABI)
+  using _Ep2 = __decay_t<_Ep>;
+  void* __ex = exception_ptr::__init_native_exception(
+      sizeof(_Ep), const_cast<std::type_info*>(&typeid(_Ep)), exception_ptr::__dest_thunk<_Ep2>);
```
Can't use `typeid` with -fno-rtti, but exceptions can be used with -fno-rtti so the previous impl would work. So this code should all be in a !_LIBCPP_HAS_NO_RTTI ifdef.

frederick-vs-ja replied:
> Can't use `typeid` with -fno-rtti, but exceptions can be used with -fno-rtti so the previous impl would work. So this code should all be in a !_LIBCPP_HAS_NO_RTTI ifdef.

Hmm... (at least on Itanium ABI?) when exceptions need to be thrown, `type_info` objects are generated and passed on throwing even if -fno-rtti is used.

Is there any intrinsic to obtain RTTI for EH when -fno-rtti is used?

jyknight reviewed (commented):
Overall, seems like a good idea; I haven't reviewed the whole thing, but here's some comments.

itrofimow committed:
revert the typeinfo header split, make exception dependent on typeinfo instead

itrofimow committed:
use weak linkage instead of dlsym

itrofimow committed:
+ clang-format

itrofimow committed:
fix transitive_includes

itrofimow committed:
some formatting again

itrofimow committed:
add libsupc++ support

itrofimow committed:
make tidy happy: + _LIBCPP_HIDE_FROM_ABI

itrofimow committed:
include new explicitly

itrofimow committed:
regenerate cxx-abilist

itrofimow committed:
drop new -> exception include dependency

itrofimow committed:
oops

itrofimow committed:
oops[2]

itrofimow committed:
fixing gcc errors hopefully

itrofimow committed:
fix windows build, add fix powerpc abilists

itrofimow committed:
fix mingw build once again

itrofimow committed:
fix abilist for freebsd, reset mingw shenanigans

itrofimow committed:
disable for apple and windows, move define to config

itrofimow committed:
+ clang-format

itrofimow committed:
fix abilists for apple

itrofimow committed:
bits of cleanup

itrofimow committed:
gcc build fix

itrofimow commented on libcxx/include/new:
```diff
@@ -370,7 +370,6 @@ inline constexpr size_t hardware_constructive_interference_size = __GCC_CONSTRUC
 _LIBCPP_END_NAMESPACE_STD
 
 #if !defined(_LIBCPP_REMOVE_TRANSITIVE_INCLUDES) && _LIBCPP_STD_VER <= 20
-#  include <exception>
```
It feels to me that this is one not-that-much-breaking change too many for one PR, but the workaround would be to split the header in public + private parts, where everything apart from these includes goes into private part.

What should be done here?

ldionne replied:
This is fine.

itrofimow reviewed (commented):
Hi! I've addressed/answered the inline comments and manually checked that libcxx builds + works with libcxxabi/libsupc++/libcxxrt (should there be a CI-build with such combinations?), so re-requesting the review.

@philnik777 @jyknight @ldionne

EricWF commented on libcxx/src/support/runtime/exception_pointer_glibcxx.ipp:
```diff
@@ -53,6 +67,36 @@ exception_ptr& exception_ptr::operator=(const exception_ptr& other) noexcept
+exception_ptr exception_ptr::__from_native_exception_pointer(void* __e) noexcept
+{
+    exception_ptr ptr{};
+    new (reinterpret_cast<void*>(&ptr)) __exception_ptr::exception_ptr(__e);
```
This has got to be an aliasing violation that the compiler can optimize on?

The types have two different sizes.

itrofimow replied:
Do they? 
https://github.com/llvm/llvm-project/blob/ea1909f82ceca30b009701f1d349247f8fbf1c3e/libcxx/src/support/runtime/exception_pointer_glibcxx.ipp#L42-L47

My understanding is that this is a hackery of course, but the hackery that should work:
https://github.com/llvm/llvm-project/blob/ea1909f82ceca30b009701f1d349247f8fbf1c3e/libcxx/src/support/runtime/exception_pointer_glibcxx.ipp#L13-L16

EricWF commented on libcxx/src/support/runtime/exception_pointer_glibcxx.ipp:
```diff
@@ -25,6 +36,9 @@ struct exception_ptr
     void* __ptr_;
 
+#  if defined(_LIBCPP_EXCEPTION_PTR_DIRECT_INIT)
+    exception_ptr(void*) noexcept;
```
Explicit please.

EricWF commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -50,15 +67,38 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
+    return exception_ptr::__from_native_exception_pointer(__ex);
   } catch (...) {
+    exception_ptr::__free_native_exception(__ex);
     return current_exception();
```
I don't understand this. The exception we're returning is the one had thrown on us during the creation of the exception pointer we wanted.

itrofimow replied:
Yes, it would be the exception thrown by an `_Ep2` copy-ctor.
As far as i know the `std::make_exception_ptr` (and `throw` as well) behaves the same: https://godbolt.org/z/o981j9oxd

EricWF commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -50,15 +67,38 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
+    } catch (...) {
+      return current_exception();
+    }
+  }
```
We should have an unreachable annotation at the bottom of this block.

itrofimow replied:
Added it, but could you please explain to me why is it needed?

EricWF commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -50,15 +67,38 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
+    ::new (__ex) _Ep2(__e);
+    return exception_ptr::__from_native_exception_pointer(__ex);
   } catch (...) {
+    exception_ptr::__free_native_exception(__ex);
```
How do we know that we don't also have to destroy the constructed exception?

itrofimow replied:
We only get here if the exception constructor threw, so there's no object to destroy. Am i missing something?

EricWF commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -50,15 +67,38 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
+#  if defined(_LIBCPP_EXCEPTION_PTR_DIRECT_INIT)
+  using _Ep2 = __decay_t<_Ep>;
+  void* __ex = exception_ptr::__init_native_exception(
+      sizeof(_Ep), const_cast<std::type_info*>(&typeid(_Ep)), exception_ptr::__dest_thunk<_Ep2>);
```
This code allocates for `_Ep` here but then constructs `_Ep2`. Similarly, it uses the typeinfo for `_Ep`.

We're sure this is correct?

ldionne commented on libcxx/include/__config:
```diff
@@ -1048,6 +1048,10 @@ __sanitizer_verify_double_ended_contiguous_container(const void*, const void*, c
 #    define _LIBCPP_HAS_NO_RTTI
 #  endif
 
+#  if !defined(_LIBCPP_HAS_NO_EXCEPTIONS) && !defined(_LIBCPP_HAS_NO_RTTI) && !defined(__APPLE__) && !defined(_WIN32)
```
Why is this disabled on Apple and on Windows? It seems to me like this should be an availability macro instead so that all platforms can take advantage of it as long as the deployment target is recent enough to support this new addition. I think this is what @philnik777 was referring to in another comment.

itrofimow replied:
The "availability macro": is that https://clang.llvm.org/docs/AttributeReference.html#availability?
Anyway, i'm failing to understand how that could eliminate the need to have a runtime check for the `__cxa_init_primary_exception` presence -- say, if i build against 18.0 libcxxabi and then LD_PRELOAD 17.0 version, which should be a supported case, right?

-------

To answer your question directly:
Windows: disabled due to mingw requiring this https://github.com/llvm/llvm-project/blob/ea1909f82ceca30b009701f1d349247f8fbf1c3e/libcxxabi/include/__cxxabi_config.h#L101 to be specified for exception destructor, and this attribute now has to appear in libc++ at the call site (where weak declaration of the `__cxa_init_primary_exception` is present). I didn't want to copy-paste this macro in libc++, but definitely doable, should i implement it?

Apple: back-deployment-macosx tests fail (https://buildkite.com/llvm-project/libcxx-ci/builds/30969#018b50e0-c82b-42e0-bceb-b74d5808ffad, for example) with `dyld[47810]: missing symbol called` error. I have very limited knowledge about how linking is performed on Apple, but my quick googling tells that it's much different from what ld does, and weak attributes don't work without some compiler flags specified. I have no clue how this should be fixed and it also takes forever for CI to reach this check, so any help would be appreciated.

itrofimow replied:
I've added a mingw support

ldionne replied:
If you look at how the availability macros in `__availability` work, basically we use the `--target <arch>-<vendor>-<platform><version>` to drive this. The idea is that you tell the compiler what's the oldest version of the platform you are targeting, and based on that we make the function available or not. If you then decide to run your program on an older platform, you effectively lied to the compiler and in practice you'll get a load-time error.

itrofimow committed:
cr fixes

itrofimow committed:
support mingw

itrofimow committed:
+ clang-format

ldionne commented on libcxx/include/__config:
```diff
@@ -1077,6 +1077,17 @@ __sanitizer_verify_double_ended_contiguous_container(const void*, const void*, c
+#    define _LIBCXX_DTOR_FUNC
+#  endif
+
+#  if !defined(_LIBCPP_HAS_NO_EXCEPTIONS) && !defined(_LIBCPP_HAS_NO_RTTI) && !defined(__APPLE__)
```
Why is this disabled on Apple?

itrofimow replied:
I changed this `ifdef` to be an availability macros instead: its' enabled when building and disabled for Apple vendoring (because the functionality is not yet released)

ldionne commented on libcxx/src/support/runtime/exception_pointer_cxxabi.ipp:
```diff
@@ -34,6 +56,37 @@ exception_ptr& exception_ptr::operator=(const exception_ptr& other) noexcept
+#  if defined(_LIBCPP_EXCEPTION_PTR_DIRECT_INIT)
+void *exception_ptr::__init_native_exception(size_t size, type_info* tinfo, void (_LIBCXX_DTOR_FUNC* dest)(void*)) noexcept
+{
+    __cxa_exception *(*cxa_init_primary_exception_fn)(void*, std::type_info*, void(_LIBCXX_DTOR_FUNC*)(void*)) = __cxa_init_primary_exception;
```
We have a usual way of dealing with newly-introduced symbols. I already explained it in my other comment about the `__availability` header. Please use that instead. There is no need for us to check whether the weak symbol exists here and do something different at runtime in case it's not. The only conditional bit we should have is in `exception_ptr.h` where we would do

```
#if _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
// use it
#else
// do the throw dance
#endif
```

itrofimow replied:
Do i understand you correctly that for libcxxrt/libsupc++ we would still need some kind of a runtime lookup?

I'm also confused about `LD_PRELOAD` case: is it okay that ld_preloading cxxabi of some previous version won't work?

ldionne replied:
Both libcxxrt and libsupc++ do implement the function, so this should not be an issue, right?

If you `LD_PRELOAD` an older version of libc++abi, then you must have provided the appropriate deployment target when compiling your program. If you do and the deployment target doesn't provide the new function, there's no issue cause the code will use the throw-dance. If you lie about the deployment target (e.g. use a recent one) but then run against an old libc++abi, you get a load-time error for the missing symbol, as you should.

itrofimow replied:
And i've adjusted that to account recent changes in `__availability`.

itrofimow committed:
cleanup runtime lookup mess, switch to an availability macros instead

itrofimow committed:
cleanup exception_pointer_glibcxx

itrofimow commented:
Hi!

I believe i've addressed most comments, with some requiring additional discussion, so could you please have a look?
Sorry for bothering you again and again, @ldionne @EricWF

itrofimow committed:
rebase fix

itrofimow committed:
remove nullptr branch, as __init_native_exception can't return nullptr anymore

ldionne commented on libcxx/include/__availability:
```diff
@@ -104,6 +104,12 @@
 #  define _LIBCPP_AVAILABILITY_BAD_ANY_CAST
 
+// These macros controls the availability of __cxa_init_primary_exception
+// in shared library, which std::make_exception_ptr might use 
```
```suggestion
// in the built library, which std::make_exception_ptr might use 
```

itrofimow replied:
Done

ldionne commented on libcxx/include/__availability:
```diff
@@ -293,4 +304,11 @@
 #   define _LIBCPP_AVAILABILITY_THROW_BAD_VARIANT_ACCESS  _LIBCPP_AVAILABILITY_BAD_VARIANT_ACCESS
 #endif
 
+// Define availability attributes that depend on both 
+// _LIBCPP_HAS_NO_EXCEPTIONS and _LIBCPP_HAS_NO_RTTI.
+#if defined(_LIBCPP_HAS_NO_EXCEPTIONS) || defined(_LIBCPP_HAS_NO_RTTI)
+#   undef _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
+#   define _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION 0
+#endif
+
```
This hunk seems unnecessary to me right now, no? We could punt on adding it and let vendors define their availability attributes correctly based on whether exceptions are enabled or not.

itrofimow replied:
The idea is to simplify the process for vendors: no matter how they define `_LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION` it just doesn't work without `RTTI` (on an unrelated note i believe it actually works with `NO_EXCEPTIONS`).

ldionne commented on libcxx/include/__config:
```diff
@@ -1023,6 +1023,13 @@ __sanitizer_verify_double_ended_contiguous_container(const void*, const void*, c
 #    define _LIBCPP_HAS_NO_RTTI
 #  endif
 
+// Same as _LIBCXXABI_DTOR_FUNC in libcxxabi, we need this when working with ABI in exceptions code
+#  if defined(_WIN32)
+#    define _LIBCXX_DTOR_FUNC __thiscall
+#  else
+#    define _LIBCXX_DTOR_FUNC
+#  endif
+
```
According to https://learn.microsoft.com/en-us/cpp/cpp/thiscall?view=msvc-170, it doesn't seem like this should be needed.

itrofimow replied:
MinGW refused to build in CI without it, but yes, lambda seems to fix the issue. Removed

ldionne commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -51,11 +71,25 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
 #  ifndef _LIBCPP_HAS_NO_EXCEPTIONS
+#    if _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
+  using _Ep2 = __decay_t<_Ep>;
+  void* __ex = exception_ptr::__init_native_exception(
+      sizeof(_Ep), const_cast<std::type_info*>(&typeid(_Ep)), exception_ptr::__dest_thunk<_Ep2>);
```
```suggestion
  void* __ex = exception_ptr::__init_native_exception(
      sizeof(_Ep), const_cast<std::type_info*>(&typeid(_Ep)), [](void* __p) {
          std::__destroy_at(static_cast<_Ep>(__p));
      });
```

Then you don't need `__dest_thunk` at all.

itrofimow replied:
Done

itrofimow replied:
Using lambda here requires C++11, so i guarded all this with `__cplusplus >= 201103L`

ldionne commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -26,6 +31,21 @@ namespace std { // purposefully not using versioning namespace
 class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
   void* __ptr_;
 
+#  if _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
```
This shouldn't be guarded by `_LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION`, only the part where you *use* `__init_native_exception` should be guarded.

itrofimow replied:
Done

ldionne commented on libcxx/lib/abi/arm64-apple-darwin.libcxxabi.v1.stable.exceptions.nonew.abilist:
```diff
@@ -2508,6 +2509,7 @@
 {'is_defined': True, 'name': '___cxa_guard_abort', 'type': 'I'}
 {'is_defined': True, 'name': '___cxa_guard_acquire', 'type': 'I'}
 {'is_defined': True, 'name': '___cxa_guard_release', 'type': 'I'}
+{'is_defined': True, 'name': '___cxa_init_primary_exception', 'type': 'I'}
```
We should be exporting `_ZNSt13exception_ptr31__from_native_exception_pointerEPv` & friends on Darwin too. You're guarding the definition of these functions in the dylib by `_LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION` but you shouldn't -- that would solve this issue.

itrofimow replied:
Done.

ldionne commented on libcxx/src/support/runtime/exception_pointer_cxxabi.ipp:
```diff
@@ -13,8 +13,9 @@
 
-exception_ptr::~exception_ptr() noexcept {
-  __cxa_decrement_exception_refcount(__ptr_);
+exception_ptr::~exception_ptr() noexcept
+{
+    __cxa_decrement_exception_refcount(__ptr_);
```
This hunk is unrelated, let's remove it.

itrofimow replied:
Done.

ldionne commented on libcxx/src/support/runtime/exception_pointer_cxxabi.ipp:
```diff
@@ -34,6 +35,29 @@ exception_ptr& exception_ptr::operator=(const exception_ptr& other) noexcept
     return *this;
 }
 
+#  if _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
```
Here, just remove the `#if` guard and it'll make sure we define these symbols whenever we build the dylib, which is what we want.

The availability macros are meant to guard whether the code in the headers should rely on these definitions in the dylib to support back-deployment use cases. But we always want to define those symbols in the dylib.

itrofimow replied:
Done.

ldionne commented on libcxx/src/support/runtime/exception_pointer_cxxabi.ipp:
```diff
@@ -34,6 +35,29 @@ exception_ptr& exception_ptr::operator=(const exception_ptr& other) noexcept
 }
 
+#  if _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
+void *exception_ptr::__init_native_exception(size_t size, type_info* tinfo, void (_LIBCXX_DTOR_FUNC* dest)(void*)) noexcept
```
Separately from the above comment, I actually wonder if we should be defining these methods in the dylib at all. Is there a reason why we're not simply calling `__cxa_allocate_exception` and `__cxa_init_primary_exception` directly from the headers?

itrofimow replied:
I thought one of the purposes of `exception_pointer_*.ipp` being in dylib is to not call `__cxa_*` from headers, but if it's not the case then we definitely can call them directly from headers.

Done just that.

ldionne commented on libcxx/src/support/runtime/exception_pointer_glibcxx.ipp:
```diff
@@ -16,6 +16,17 @@
 // function.
 
+
+#  if _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
```
I don't understand why we're modifying this file in this PR. We added a new function to libc++abi, we didn't touch libsupc++ though.

itrofimow replied:
Although we didn't change libsupc++, we added some code on the call site which invokes `__cxa_init_primary_exception`.
Since the function is present in libsupc++ i thought it would make sense to support the optimization for libsupc++ as well, and that requires some changes in this file.

I'm totally fine with not doing it in this PR

ldionne commented on libcxxabi/include/cxxabi.h:
```diff
@@ -36,13 +36,19 @@ class type_info; // forward declaration
 __cxa_allocate_exception(size_t thrown_size) throw();
 extern _LIBCXXABI_FUNC_VIS void
 __cxa_free_exception(void *thrown_exception) throw();
+// This function is an LLVM extension, which mirros the same extension in libsupc++ and libcxxrt
```
```suggestion
// This function is an LLVM extension, which mirrors the same extension in libsupc++ and libcxxrt
```

itrofimow replied:
Done.

ldionne reviewed (changes requested):
Thank you, this is looking a whole lot better now! I have some comments and questions, but I think this will be able to move forward once everything has been addressed.

itrofimow committed:
cr fixes, simplify: call __cxa_* directly from headers

itrofimow committed:
fix abi-list for unknown-linux-gnu

itrofimow committed:
fix the build for C++03

itrofimow committed:
fix the build for C++03, attempt 2

itrofimow committed:
fix transitive includes for cxx03

itrofimow committed:
fix transitive includes for cxx11,14,17,20

itrofimow committed:
adjust abi-lists for other platforms

itrofimow committed:
adjust abi-list for freebsd

itrofimow committed:
adjust the abi-list for noexceptions

itrofimow committed:
merge with upstream

itrofimow committed:
an attempt to fix MinGW build

itrofimow committed:
an attempt to fix MinGW build once again

itrofimow commented:
I've addressed/answered inline comments and the build is finally green.

I'm not particularly in love with declaring `__cxa_allocate_exception`, `__cxa_free_exception` and `__cxa_init_primary_exception` in `exception_ptr.h`, but i find it's better than including whole `cxxabi.h` there.

Could you please have a look? @ldionne

itrofimow committed:
merge with upstream

itrofimow committed:
cleanup redundant whitespaces

itrofimow committed:
merge fix

itrofimow committed:
don't declare cxa_* functions for Microsoft ABI

itrofimow committed:
formatting fix

itrofimow committed:
Merge remote-tracking branch 'upstream/main' into dont_throw_in_make_exception_ptr

itrofimow committed:
Merge remote-tracking branch 'upstream/main' into dont_throw_in_make_exception_ptr

itrofimow committed:
Merge remote-tracking branch 'upstream/main' into dont_throw_in_make_exception_ptr

itrofimow committed:
update abilists for Android

ldionne commented on libcxx/include/__availability:
```diff
@@ -299,4 +310,11 @@
+// Define availability attributes that depend on both
+// _LIBCPP_HAS_NO_EXCEPTIONS and _LIBCPP_HAS_NO_RTTI.
+#if defined(_LIBCPP_HAS_NO_EXCEPTIONS) || defined(_LIBCPP_HAS_NO_RTTI)
+#   undef _LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION
```
You also need to `#undef _LIBCPP_AVAILABILITY_INIT_PRIMARY_EXCEPTION` and `#define _LIBCPP_AVAILABILITY_INIT_PRIMARY_EXCEPTION` to make that macro a no-op when RTTI / exceptions are disabled.

ldionne commented on libcxx/src/support/runtime/exception_pointer_glibcxx.ipp:
```diff
@@ -23,6 +23,7 @@ namespace __exception_ptr {
 struct exception_ptr {
   void* __ptr_;
 
+  explicit exception_ptr(void*) noexcept;
```
Where is that function defined?

itrofimow replied:
AFAIU all these functions are defined in libsupc++, this one is defined here: https://github.com/gcc-mirror/gcc/blob/master/libstdc%2B%2B-v3/libsupc%2B%2B/eh_ptr.cc#L75

ldionne commented on libcxx/include/__exception/exception_ptr.h:
```diff
@@ -9,23 +9,57 @@
+    std::type_info*,
+    void(
+#  if defined(_WIN32)
+        __thiscall
```
I'm a bit confused -- why do we need this `#if defined(_WIN32)` and `__thiscall` if this whole block is disabled on Microsoft via `#ifndef _LIBCPP_ABI_MICROSOFT` above?

itrofimow replied:
MinGW requires this: https://buildkite.com/llvm-project/libcxx-ci/builds/32388#018c84de-f231-43b4-839a-c9fe88f118fe/311-1557

This is how `__cxa_throw` (and `__cxa_init_primary_exception` as well) is defined: https://github.com/llvm/llvm-project/blob/2bb2a42fa261d728e32ba0b1f9cf27ba7991440f/libcxxabi/include/cxxabi.h#L48-L55
 
and `_LIBCXXABI_DTOR_FUNC` expands to `__thiscall` under win32: https://github.com/llvm/llvm-project/blob/2bb2a42fa261d728e32ba0b1f9cf27ba7991440f/libcxxabi/include/__cxxabi_config.h#L100-L101

itrofimow replied:
P.S. `__USING_WASM_EXCEPTIONS__` usage here confuses me: it's not used consistently across the libcxxabi, and i'm not really sure what this is about and whether I should adjust the signature of `__cxa_init_primary_exception`.
I assume i should, right?

ldionne replied:
This looks OK to me. The wasm folks will reach out if they need to do something special here.

ldionne reviewed (commented):
A few minor comments. I'll want to take a last look before shipping but I think this is looking really good now.

itrofimow committed:
Merge remote-tracking branch 'upstream/main' into dont_throw_in_make_exception_ptr

itrofimow committed:
undef _LIBCPP_AVAILABILITY_INIT_PRIMARY_EXCEPTION under _LIBCPP_HAS_NO_EXCEPTIONS or _LIBCPP_HAS_NO_RTTI

itrofimow committed:
remove unused include

itrofimow committed:
fix the docs for building libcxx with other ABI-libraries

itrofimow commented:
I had to spend some time bisecting very cryptic `error: undefined symbol: _Unwind_Resume` when building libc++ with libcxxrt/libsupc++ to this PR https://github.com/llvm/llvm-project/pull/77687, so i've adjusted the docs to explicitly disable flipped default.

Building with libcxxrt also requires `DLIBCXX_ENABLE_NEW_DELETE_DEFINITIONS` to be set, otherwise `operator new`/`operator delete` are undefined.

Also, addressing the last outstanding comment about `undef`-ing the availability macro i realized that `_LIBCPP_AVAILABILITY_INIT_PRIMARY_EXCEPTION` isn't used anywhere and I'm not really sure to what it should be applied.

ldionne reviewed (approved):
Thanks, LGTM!

itrofimow commented:
The tests failing are ARMv7-picolib tests, and they fail with
```
Unexpectedly Passed Tests (2):
  armv7m-picolibc-libc++.cfg.in :: libcxx/selftest/modules/std-module.sh.cpp
  armv7m-picolibc-libc++.cfg.in :: std/modules/std.pass.cpp
```
which i don't think is related to this PR changes

ldionne commented:
Indeed, those CI failures are unrelated and should be fixed on `main`. Merging this now.

itrofimow commented:
> Indeed, those CI failures are unrelated and should be fixed on `main`. Merging this now.

Thank you for all the time you've spent helping me with this PR.
Please don't hesitate to ping me here or on discord if something goes wrong with these changes.

itrofimow commented:
Hi @nico! No, that is absolutely not intentional, i've overlooked that, sorry.
Could you please check if your bot builds with this patch applied? https://github.com/llvm/llvm-project/pull/79040/files

nico commented:
I checked out 309dcc112b1d3f0a24ea67e5b0377c649637260f^ and verified that I could repro the link error. Then I patched in https://github.com/llvm/llvm-project/pull/79040 and verified that it went away. So that PR helps, thanks!

DimitryAndric commented:
FWIW this caused an incompatibility with recent libcxxrt (after https://github.com/libcxxrt/libcxxrt/commit/03c83f5a57be8c5b1a29a68de5638744f17d28ba), since its `cxxabi.h` declares `__cxa_allocate_exception`, `__cxa_free_exception` and `__cxa_init_primary_exception` without `throw()` specifiers.

I submitted https://github.com/libcxxrt/libcxxrt/issues/25, since I am not sure about the official signature of these functions, if there is any.

ldionne commented:
@DimitryAndric Thanks, please keep us in the loop. If we need to make a fix on our side, @itrofimow we'll want to cherry-pick it to LLVM 18.

dankm commented:
@ldionne it looks like the consensus is that GCC and LLVM are correct and that we're fixing libcxxrt. The relevant pull request is libcxxrt/libcxxrt#27.

ldionne commented:
Thanks for the heads up!

itrofimow commented:
I'm looking into it
